README.md
# Parallel Editor

> Edit component’s parts(tamplate/script/style) use 3 columns view, It’s layout looks like CodePen/JSFiddle/JSBin, used for Angular components, Vue TypeScript components or other mvc/mvvm components which itself’s templates/scripts/styles parts each as single file wrappered with in the same folder.

---

## Features

1.  Automatically open scripts, templates, style files of the same component in multipe columns;
2.  Supported a variety of template, style, script extnames;

    > * _Scripts_: `.js`, `.jsx`, `.ts`, `.tsx`, `.mjs`, `.es`, `.es6`, `.coffee`, `.dart`
    > * _Styles_: `.css`, `.scss`, `.sass`, `.less`, `.styl`, `.stylus`
    > * _Templates_: `.html`, `.htm`, `.xhtml`, `.ejs`, `.jade`, `.pug`, `.hbs`, `.handlebars`, `.haml`, `.tpl`, `.mustache`, `.def`, `.dot`, `.jst`, `.dust`, `.njk`

3.  Automatically identifies the component's constituent files.
4.  Support to adjust the order of `scripts`, `templates`, and `style` columns in `Extension Settings`.
5.  When we cannot identify which files belong to the same component, we will ask you to confirm the correct file that makes up the component and remember your choices.
6.  Remember component's part files status.
7.  (developing) Auto split `single-file component` parts into each column on editting, and merged they are together on save.

---

### Parallel How to detecting part files of components?

1.  Same named files will be treated as a set of files for the same component.
    ```
    ┣━ Components
      ├┈ index.ts
      ├┈ common.ts
      ├┈ ...
      ├┈ base-button.ts       ↘
      ├┈ base-button.scss     → BaseButtom component
      ├┈ base-button.html     ↗
      ├┈ ...
      ├┈ Tab.ts               ↘
      ├┈ tab.scss             → Tab component (Case insensitive)
      └┈ tab.html             ↗
    ```
1.  In the folder only contains one script file, one style file, one template file will be treated as a component.

    ```
    ┣━ Components
      ├┈ index.ts
      └┈ Icon
        ├┈ index.ts           ↘
        ├┈ icon.html          → Icon component
        └┈ icon.scss          ↗
      └┈ Avatar
        ├┈ index.tsx          ↘
        └┈ style.scss         → Icon component (no single template file, it's included in .tsx use JSX format.)
      └┈ Foo
        └┈ ...
      └┈ Bar
        └┈ ...
    ```

---

## Settings

### How to find `Parallel Editor Settings`?

1.  Open `VSCode Settings`
2.  Find `Parallel Editor` section or search for `parallel.` prefix.

### Parallel Editor Settings:

##### - `parallel.cacheDirectory`

* **type:** `string`
* **default:** `.parallel`
* **description:** A directory under the current VSCode Workspace use for saving Parallel Editor's cache files, temp files.

##### - `parallel.scriptExtnames`

* **type:** `array`
* **default:** `[]`
* **built-in:** `['.js', '.jsx', '.ts', '.tsx', '.mjs', '.es', '.es6', '.coffee', '.dart']`
* **description:** A set of extnames used for detect component's script file.
  > **Notes:**
  >
  > * The extnames you setted will be merged into the built-ins and deduplicated.
  >
  > * if you want to exclude built-in script extnames, please use `!` prefix symbol.
  >   <br/><br/> **For example:**
  >
  >   ```javascript
  >   {
  >     // ...
  >     parallel.scriptExtnames: [
  >       '!.ts', // excluded built-in .ts extname
  >       '!.tsx', // excluded built-in .tsx extname
  >       '.component.ts', // includes
  >     ],
  >     // ...
  >   }
  >   ```
  >
  >   **matched results:**
  >
  >   * [x] `app.component.ts` **(matched)**
  >   * [ ] `app.module.ts`
  >   * [ ] `app.ts`

##### - `parallel.styleExtnames`

* **type:** `array`
* **default:** `[]`
* **built-in:** `['.css', '.scss', '.sass', '.less', '.styl', '.stylus' ]`
* **description:** A set of extnames used for detect component's style file.

  > **Notes:**
  >
  > * The extnames you setted will be merged into the built-ins and deduplicated.
  >
  > * if you want to exclude built-in style extnames, please use `!` prefix symbol.
  >   <br/><br/> **For example:**
  >
  >   ```javascript
  >   {
  >     // ...
  >     parallel.styleExtnames: [
  >       '!.sass', // excluded built-in .sass extname
  >       '.component.sass', // includes `*.component.sass`
  >     ],
  >     // ...
  >   }
  >   ```
  >
  >   **matched results:**
  >
  >   * [x] `app.component.sass` **(matched)**
  >   * [ ] `app.module.sass`
  >   * [ ] `app.sass`

##### - `parallel.templateExtnames`

* **type:** `array`
* **default:** `[]`
* **built-in:** `['.css', '.scss', '.sass', '.less', '.styl', '.stylus' ]`
* **description:** A set of extnames used for detect component's template file.

  > **Notes:**
  >
  > * The extnames you setted will be merged into the built-ins and deduplicated.
  >
  > * if you want to exclude built-in template extnames, please use `!` prefix symbol.
  >   <br/><br/> **For example:**
  >
  >   ```javascript
  >   {
  >     // ...
  >     parallel.templateExtnames: [
  >       '!.html', // excluded built-in .html extname
  >       '.component.html', // includes `*.component.html`
  >     ],
  >     // ...
  >   }
  >   ```
  >
  >   **matched results:**
  >
  >   * [x] `app.component.html` **(matched)**
  >   * [ ] `app.module.html`
  >   * [ ] `app.html`

##### - `parallel.includeDirectories`

* **type:** `array`
* **default:** `[]`
* **built-in:** `['/components', '/views', '/pages', '/packages']`
* **description:** enable Parallel Editor for those paths and sub paths.

  > **Notes:**
  >
  > The dirs you setted will be merged into the built-ins and deduplicated.

##### - `parallel.columnsOrder`

* **type:** `array`
* **default:** `['script', 'template', 'style']`
* **description:** Set script, template, style panels layout order in the editor.

##### - `(TODO)parallel.splitSingleFileComponentOnEditing`

* **type:** `boolean`
* **default:** `true`
* **description:** whether auto split templates/scripts/style tags in multiple columns view for a single-file component when open it.

##### - `(TODO)parallel.singleFileComponentExtnames`

* **type:** `array`
* **default:** `[]`
* **built-in:** `['.vue', '.we', '.weex']`
* **description:** enable split mode for Single-file Component which with those file extnames.

## Todos

* [ ] Support Single-file Component split editing mode.
* [ ] Detect components parts cross multiple directories.
* [ ] Generate components.

---

### For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2018-04-17

### Added

* Supported automaticly open the component's parts such as script file, template file, style file in each VS Code editor column.
* Supported multiple file type detections for component parts.
* Supported adjust columns order in Settings.
* Supported add custom extnames in Settings.
* Supported components filepath matching patterns in Settings.

### Changed

* none.

### Removed

* none.


src/config.ts
/*
 * @Author: Colin Luo
 * @Date: 2018-04-17 06:30:34
 * @Last Modified by: Colin Luo <mail@luozhihua.com>
 * @Last Modified time: 2018-04-22 04:34:31
 */
import { workspace } from 'vscode';
// import * as mm from 'micromatch';
// import * as walker from 'klaw-sync';

type RA = ReadonlyArray<string>;
export interface Exts {
  [index: string]: RA;
  style: RA;
  script: RA;
  template: RA;
}

const DEF_SCRIPT_DIR: RA = [
  'script',
  'controller',
  'ctrl',
  'javascript',
  'typescript',
  'coffeescript',
];
const DEF_STYLE_DIR: RA = ['style'];
const DEF_TEMPLATE_DIR: RA = ['template', 'view', 'page'];
const DEF_COMPONENT_DIR: RA = ['component', 'view', 'page', 'src/app'];
const DEF_SFC_EXTS: RA = ['.vue', '.we', '.weex'];
const DEF_COL_ORDER: RA = ['script', 'template', 'style'];
const DEF_EXTS: Exts = {
  style: ['.css', '.scss', '.sass', '.less', '.styl', '.stylus'],
  script: [
    '.js',
    '.jsx',
    '.ts',
    '.tsx',
    '.mjs',
    '.es',
    '.es6',
    '.coffee',
    '.dart',
  ],
  template: [
    '.jade',
    '.pug',
    '.tpl',
    '.html',
    '.mustache',
    '.ejs',
    '.def',
    '.dot',
    '.jst',
    '.handlebars',
    '.hbs',
    '.haml',
    '.dust',
    '.njk',
  ],
};

export default class Config {
  constructor() {}

  public readonly defaultColumnsOrder: RA = DEF_COL_ORDER;
  public readonly defaultExts: Exts = DEF_EXTS;
  public readonly defaultScriptDir: RA = DEF_SCRIPT_DIR;
  public readonly defaultStyleDir: RA = DEF_STYLE_DIR;
  public readonly defaultTemplateDir: RA = DEF_TEMPLATE_DIR;
  public readonly defaultComponentDir: RA = DEF_COMPONENT_DIR;
  public readonly defaultSFCExts: RA = DEF_SFC_EXTS;

  public get exts(): Exts {
    let extnames: Exts = {
      style: this.styleExts,
      script: this.scriptExts,
      template: this.templateExts,
    };

    return extnames;
  }

  public get scriptDirs() {
    let { scriptFoldersForCrossMode } = this.getWorkspaceConfig();
    let exts: string[] = this.scriptExts.reduce((res, cur) => {
      return res.concat(cur.substr(1) as never);
    }, []);

    return this.mergePatterns(DEF_SCRIPT_DIR, [
      ...scriptFoldersForCrossMode,
      ...exts,
    ]);
  }

  public get styleDirs() {
    let { styleFoldersForCrossMode } = this.getWorkspaceConfig();
    let exts: string[] = this.styleExts.reduce((res, cur) => {
      return res.concat(cur.substr(1) as never);
    }, []);

    return this.mergePatterns(DEF_STYLE_DIR, [
      ...styleFoldersForCrossMode,
      ...exts,
    ]);
  }

  public get templateDirs() {
    let { templateFoldersForCrossMode } = this.getWorkspaceConfig();
    let exts: string[] = this.templateExts.reduce((res, cur) => {
      return res.concat(cur.substr(1) as never);
    }, []);

    return this.mergePatterns(DEF_TEMPLATE_DIR, [
      ...templateFoldersForCrossMode,
      ...exts,
    ]);
  }

  public get componentDirs() {
    let { componentFolders } = this.getWorkspaceConfig();

    return this.mergePatterns(DEF_COMPONENT_DIR, componentFolders);
  }

  public get styleExts(): string[] {
    let { styleExts } = this.getWorkspaceConfig();

    return this.mergePatterns(DEF_EXTS.style as string[], styleExts);
    // return this.getStyleExts();
  }

  public get scriptExts(): string[] {
    let { scriptExts } = this.getWorkspaceConfig();

    return this.mergePatterns(DEF_EXTS.script as string[], scriptExts);
  }

  public get templateExts(): string[] {
    let { templateExts } = this.getWorkspaceConfig();

    return this.mergePatterns(DEF_EXTS.template as string[], templateExts);
  }

  public get isSplitSFC(): boolean {
    let { splitSingleFileComponentOnEditing } = this.getWorkspaceConfig();

    return !!splitSingleFileComponentOnEditing;
  }

  public get sfcExts(): string[] {
    let { singleFileComponentExts } = this.getWorkspaceConfig();

    return this.mergePatterns(
      DEF_SFC_EXTS as string[],
      singleFileComponentExts,
    );
  }

  public get columnsOrder(): string[] {
    let { columnsOrder } = this.getWorkspaceConfig();
    let defaultOrder = config.defaultColumnsOrder;
    let order = [...columnsOrder];

    defaultOrder.forEach(item => {
      if (!columnsOrder.includes(item)) {
        order.push(item);
      }
    });

    return order;
  }

  public getWorkspaceConfig(): any {
    return workspace.getConfiguration('parallel');
  }

  /**
   * @description 对匹配规则去重和排除
   * @static
   * @param {string[]} defaults
   * @param {string[]} patterns
   * @returns {string[]}
   * @memberof Config
   */
  public mergePatterns(
    defaults: string[] | RA,
    patterns: string[] | RA,
  ): string[] {
    let allPatterns: string[];

    // 是否排除所有内置匹配规则
    if (patterns.includes('!built-ins')) {
      allPatterns = patterns as string[];
    } else {
      allPatterns = [...defaults, ...patterns];
    }

    return allPatterns
      .sort() // put ! ahead
      .reverse() // put ! at the last
      .reduce((cleared, pattern) => {
        if (pattern.startsWith('!')) {
          pattern = pattern.substr(1);
          return cleared.filter(item => item !== pattern);
        } else {
          return cleared
            .filter(item => item !== pattern)
            .concat(pattern as never);
        }
      }, []);
  }
}

export let config: Config = new Config();
